### ISSUE TITLE
coursier ignore artifacts (jar files) with packaging=pom

### ISSUE THREAD

**Issue by rumoku** [alexarchambault/coursier#278]
As an example zookeeper artifact published with packagin equals pom.
Coursier as the result fetch pom only, while sbt (and maven) fetch both pom+jar.
What can be done with that?

> $ ./coursier fetch org.apache.zookeeper:zookeeper:3.4.6
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\junit\junit\3.8.1\junit-3.8.1.jar
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\org\apache\zookeeper\zookeeper\3.4.6\zookeeper-3.4.6.pom
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\jline\jline\0.9.94\jline-0.9.94.jar
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\org\slf4j\slf4j-api\1.6.1\slf4j-api-1.6.1.jar
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\org\slf4j\slf4j-log4j12\1.6.1\slf4j-log4j12-1.6.1.jar
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\io\netty\netty\3.7.0.Final\netty-3.7.0.Final.jar
> C:\Users\Vladimir\.coursier\cache\v1\https\repo1.maven.org\maven2\log4j\log4j\1.2.16\log4j-1.2.16.jar
> 

Quote above shows that eventually coursier fetch all *jar dependencies of zookeeper, but ignore this jar itself.

I've also created sample project for investigation: https://github.com/rumoku/coursier-zoo-issue
The funny thing there: if zookeeper added as transitive depdendency it's fetched perfectly :\

**Comment by alexarchambault** [comment]
My current opinion is that the "working" case in coursier is an accident, and the pom packaging in zookeeper is one too...

The JAR being fetched when it's a transitive dependency must originate from early attempts to have hadoop / spark-related dependencies work fine (these typically pull zookeeper).

There are dependencies with `pom` packaging that actually have no JARs (e.g. `com.github.fommil.netlib:all:1.1.2`). So not trying to fetch a JAR seems legit. And the `pom` packaging in zookeeper totally looks like an accident :-) zookeeper is apparently built with Ivy, and tries to set a `packaging` parameter to `jar` [here](https://github.com/apache/zookeeper/blob/release-3.4.6/build.xml#L1136). Reading [the Ivy doc](http://ant.apache.org/ivy/history/2.4.0/use/makepom.html), it might not be the right way to tell the packaging to Ivy, and Ivy then defaults to... `pom`.

Fixing the accident in coursier is likely to break spark et al. Defaulting to the JAR type for manually added dependencies would break SBT projects adding directly `com.github.fommil.netlib:all:1.1.2` (`packaging` was taken into to handle it IIRC).

**Comment by alexarchambault** [comment]
I'm not sure what would be the right thing to do here... Maybe fixing the "accident" in coursier, and having a hard-coded list of dependencies whose packaging would be forced to JAR by default. With options to disable reading this hard-coded list.

**Comment by rumoku** [comment]
@alexarchambault, correct me if I'm wrong. 
My assumption that both sbt (with default ivy resolver) and maven are looking into classifier that we register together with dependency with defaulting to jar if missed - and perform relevant fetching.
Is that sounds reasonable?

So if I want to load jar, I simply add dependency:
libraryDependencies += "org.apache.zookeeper" % "zookeeper" % "3.4.8"

But if we want to fetch pom only in sbt I suppose to do smth like:
`libraryDependency += "com.github.fommil.netlib" % "all" % "1.1.2"  pomOnly()`

**Comment by rumoku** [comment]
As I show in test project: sbt without coursier plugin and scala are fetching zookeeper jar (directly).

**Comment by cheleb** [comment]
As an illustration, this project suffer of the issue (zookeeper fault ...) https://github.com/twitter/util

**Comment by cheleb** [comment]
A simple (temporary) fix for the issue, that copies zookeeper in unmanaged "lib" folder:

```
val zookeeperVersion = "3.4.8"
val fixZoo : State => State = {
   s=>
  //println(s.attributes.get("coursierSbtResolvers"))
  val fileName = s"zookeeper-${zookeeperVersion}.jar"
   val e = new File( s"lib/$fileName")
  if(!e.exists())
   sbt.IO.download(new URL(s"http://repo1.maven.org/maven2/org/apache/zookeeper/zookeeper/$zookeeperVersion/$fileName"), e)
     
 s
  
}
onLoad in Global ~= (fixZoo compose _)	
```

**Comment by alexarchambault** [comment]
@cheleb https://github.com/alexarchambault/coursier/pull/290 should fix that.

Another workaround would be to manually ask for the JAR via
```scala
libraryDependencies += "org.apache.zookeeper" % "zookeeper" % "3.5.0-alpha" artifacts(Artifact(""))
```

**Comment by cheleb** [comment]
Much simpler ! Thank a lot.

**Comment by pdex** [comment]
> I'm not sure what would be the right thing to do here... Maybe fixing the "accident" in coursier, and having a hard-coded list of malformed dependencies whose packaging would be forced to JAR by default. With options to disable reading this hard-coded list.

Maybe with more context the solution is easier to understand? If you're performing a fetch operation that expects a JAR you should be returning a .jar and not a .pom

**Comment by pdex** [comment]
I'm also confused as to why coursier isn't using the same resolution logic as ivy or maven.

**Comment by lins05** [comment]
> Fixing the accident in coursier is likely to break spark et al.

@alexarchambault When building spark from source with hive support (e.g. `cd spark; ./build/sbt -Phive package`), the compile would fail if sbt coursier plugin is used (I use 1.0.0-M15).

The reason is spark's hive support depends on `org.apache.thrift:libfb303:0.9.3`, which is packaged with type `pom` as well. 

I tried each of:

1) copy the libfb303 jar to the proper folder under the coursier cache directory,  and 
2) add `libraryDependencies += "org.apache.thrift" % "libfb303" % "0.9.3" jar()` in `local.sbt` under spark source root. 

But neither of them works. So I have to comment out the sbt coursier plugin to get around this.

**Comment by avdv** [comment]
Hi.

**TL;DR** setting `coursierDefaultArtifactType` to `"jar"`, makes this work for me again. (this has been the default in #290, but was reverted again to fix #322)

It took me a while to figure this out while I wrote up the following text. Maybe it turns out to be useful for others searching for a solution that's why I am keeping it here.

---

I also ran into this problem again when trying to run `testUtils/compile` in scalafmt which has a dependency on commons-io:

```
"org.apache.commons" % "commons-io"  % "1.3.2"
```

I have set `useJCenter := true` in my `~/.sbt/0.13/global.sbt` build file.

```
> show testUtils/update
[info] Update report:
[info] 	Resolve time: -1 ms, Download time: -1 ms, Download size: -1 bytes
...
[info] 	compile:
[info] 		org.apache.commons:commons-io:1.3.2:default: (Artifact(commons-io,pom,pom,None,List(),Some(https://jcenter.bintray.com/org/apache/commons/commons-io/1.3.2/commons-io-1.3.2.pom),Map()),/home/claudio/.coursier/cache/v1/https/jcenter.bintray.com/org/apache/commons/commons-io/1.3.2/commons-io-1.3.2.pom)
...
```

Apparently, JCenter has a different POM file than Maven Central:

* https://jcenter.bintray.com/org/apache/commons/commons-io/1.3.2/commons-io-1.3.2.pom
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
<modelVersion>4.0.0</modelVersion>
<parent>
<artifactId>commons-parent</artifactId>
<groupId>org.apache.commons</groupId>
<version>3</version>
</parent>
<groupId>org.apache.commons</groupId>
<artifactId>commons-io</artifactId>
<version>1.3.2</version>
<packaging>pom</packaging>
<name>Commons IO</name>
...
```
* http://central.maven.org/maven2/org/apache/commons/commons-io/1.3.2/commons-io-1.3.2.pom
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
<modelVersion>4.0.0</modelVersion>
<groupId>org.apache.commons</groupId>
<artifactId>commons-io</artifactId>
<version>1.3.2</version>
<distributionManagement>
<relocation>...</relocation>
</distributionManagement>
</project>
```

**Comment by volth** [comment]
How did you end up separating ```pom```s with ```<packaging>pom</packaging>``` like 
http://repo1.maven.org/maven2/io/monix/monix_2.11/2.3.0/monix_2.11-2.3.0.pom (which have no .jar) http://repo1.maven.org/maven2/org/apache/zookeeper/zookeeper/3.4.9/zookeeper-3.4.9.pom (which have a .jar).
Coursier works with it now, but what is the magic?

**Comment by alexarchambault** [comment]
@volth Since `1.0.0-RC1`, coursier looks at the directory listings (e.g. `http://repo1.maven.org/maven2/io/monix/monix_2.11/2.3.0/`), to know which artifacts are available, rather than relying solely on the packaging tag. This works fine for most public repositories, but doesn't for Nexus proxies, which either don't allow directory listings, or only list what they copied locally... So since `1.0.0-RC4`, coursier also "tries" to download JARs, allowing them to be not found (the not found error being ignored if the `optional` method on `Artifact` is true). Overall, the packaging tag is now only used to set the "type" of the jar, if it's fine for a JAR (e.g. `bundle`).
